### ISSUE TITLE
MultiplexingStream significantly slower than raw stream access

### ISSUE THREAD

**Issue by AArnott** [AArnott/Nerdbank.Streams#505]
### Discussed in https://github.com/AArnott/Nerdbank.Streams/discussions/503

<div type='discussions-op-text'>

<sup>Originally posted by **plonkaadam** July  1, 2022</sup>
I'm trying to send a large binary file using `MultiplexingStream` over `TcpSocket`. I tried to send it directly using sockets (for now on loopback interface) and I'm able to transfer over 10GB/s.
When I try to do the same using `MultiplexingStream` and return `Stream`, `PipeReader` or create separate channel I could get up to only 80MB/s.

I tried to increase the buffer sizes but with no luck. Does the `MultiplexingStream` add so much overhead that it's not possible to get it any faster?

// after lot of tries I found out that there is a significant difference in speed between versions and I tried to use the newest one
Average speeds I got on my laptop are:
V1 - 118.11 MBps
V2 - 41.5 MBps
V3 - 48.0 MBps

Anyway - this application will run on a much slower device which during my initial tests got only up to 8MBps on V3; there's still a huge gap between multiplexed stream and raw stream speeds. Is there any way to make it faster?
Great library btw</div>

In testing this, I measured 2-4Gbps using FullDuplexStream (basically raw pass through with no I/O overhead). That represents the ceiling of perf we might achieve (on my machine). MultiplexingStream V1 achieves 1.3Gbps, V2 931Mbps, and V3 599Mbps. So we indeed have a perf hit, and it gets worse with each protocol version. MessagePack slows down v2, and I'm not sure about V3 (window size handling, perhaps?). 
What's particularly interesting is that when the transport is a named pipe instead of FullDuplexStream, mxstream slows to below 100Mbps. We should compare raw binary transfers across named pipes, for a fair comparison. But it may be that the frequency of flushes or some other Stream call has particularly high overhead when doing _real_ I/O that involves the Windows kernel. We should look into both.

**Comment by AArnott** [comment]
I wonder if #532 being fixed will alleviate this somewhat.

**Comment by adamplonka** [comment]
I digged a little into this and seems that increasing `FramePayloadMaxLength` greatly improves the performance. To do that I also had to increase the maximum array size in `SequencePool.arrayPool`.
Changing it from 20480 to 65535 bytes sped up the stream transfer from 3.1 seconds down to 1.2 seconds on V1 and from average of 7.1 seconds down to 2.6 seconds on V3.
It's not trivial to increase the payload size beyond 65535 bytes in V1 and V2 because there are only two bytes reserved for payload size in the frame header.
Increasing the payload size even further is possible in V3 where I got down to just 0.9 seconds with a 1MB buffer. Using 256kB buffer makes V3 as fast as V1.

Can you make the frame payload length configurable and let it be more than 64kB? For backward compatibility you probably cannot change the frame header format in V1 but can there be something like V1.1?
It would be useful for my use cases where I transfer files within the same local network and do local rpc calls.

**Comment by Elgot** [comment]
Is there any progress on this?

**Comment by AArnott** [comment]
No, I'm afraid it hasn't been a priority for my use cases. 

Sponsorship may be an effective way to lead a contributor of this repo to prioritize the feature. As a free and open source library, all or the majority of its development is done without any monetary benefit to its authors, and all of it is made available free of charge. But sponsorship is very much appreciated and, with agreement from a particular contributor and/or the owning team, can be a great way to support the project and help get your favorite features or bugs addressed in an upcoming version.
If sponsorship to drive completion of a feature or bug fix is an interesting option to you, please let us know. You can reach out to me privately on [keybase.io](https://keybase.io/aarnott) or [gitter.im](https://gitter.im/aarnott) as well.

**Comment by Elgot** [comment]
Setting _BackpressureSupportEnabled_ to _false_ seems to improve the throughput 2-3 times, so that is perhaps the main difference between V1 and V3.

**Comment by Elgot** [comment]
Actually the real world (or "Start without debugging" in VS) performance seems much better (typically 1 GB/s vs 40 MB/s in VS). Still three times slower than FullDuplexStream, but probably good enough for most cases.

**Comment by adamplonka** [comment]
@Elgot yes I noticed the same, increasing the maximum message size to 64kB (maximum allowed by v1 of protocol) improves the performance greatly in debug mode but has no real impact to performance when running without debugger
